core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+import static java.util.Objects.requireNonNull;
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> finishedTasks = new HashSet<>();
+    @GuardedBy("this")
+    private final Map<TaskId, Throwable> failedTasks = new HashMap<>();
+    @GuardedBy("this")
+    private boolean inputFinished;
+    @GuardedBy("this")
+    private Throwable failure;
+
+    @GuardedBy("this")
+    private final ListMultimap<TaskId, SerializedPage> pageBuffer = ArrayListMultimap.create();
+    @GuardedBy("this")
+    private Iterator<SerializedPage> pagesIterator;
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+
+    @GuardedBy("this")
+    private boolean closed;
+
+    public DeduplicationExchangeClientBuffer(Executor executor, DataSize bufferCapacity, RetryPolicy retryPolicy)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+        this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    @Override
+    public synchronized SerializedPage pollPage()
+    {
+        throwIfFailed();
+
+        if (closed) {
+            return null;
+        }
+
+        if (!inputFinished) {
+            return null;
+        }
+
+        if (pagesIterator == null) {
+            pagesIterator = pageBuffer.values().iterator();
+        }
+
+        if (!pagesIterator.hasNext()) {
+            return null;
+        }
+
+        SerializedPage page = pagesIterator.next();
+        pagesIterator.remove();
+        bufferRetainedSizeInBytes -= page.getRetainedSizeInBytes();
+
+        return page;
+    }
+
+    @Override
+    public synchronized void addTask(TaskId taskId)
+    {
+        if (closed) {
+            return;
+        }
+
+        checkState(!noMoreTasks, "no more tasks expected");
+        checkState(allTasks.add(taskId), "task already registered: %s", taskId);
+
+        if (retryPolicy == QUERY) {
+            removePagesForPreviousAttempts(taskId.getAttemptId());
+        }
+    }
+
+    @Override
+    public synchronized void addPages(TaskId taskId, List<SerializedPage> pages)
+    {
+        if (closed) {
+            return;
+        }
+
+        checkState(allTasks.contains(taskId), "task is not registered: %s", taskId);
```

Comment:
should you ignore the call if `retryMode == QUERY` and there are already pages for newer attempt for given task?

Reply:
Yeah, the pages should be ignored. Let me add a check